src/app/devtools/networklogger/qgsnetworkloggerpanelwidget.cpp

```diff
@@ -0,0 +1,144 @@
+{
+  mProxyModel->setShowSuccessful( show );
+}
+
+void QgsNetworkLoggerTreeView::setShowTimeouts( bool show )
+{
+  mProxyModel->setShowTimeouts( show );
+}
+
+void QgsNetworkLoggerTreeView::itemExpanded( const QModelIndex &index )
+{
+  // if the item is a QgsNetworkLoggerRequestGroup item, open all children (show ALL info of it)
+  // we want to scroll to last request
+
+  // only expand all children on QgsNetworkLoggerRequestGroup nodes (which don't have a valid parent!)
+  if ( !index.parent().isValid() )
+    expandChildren( index );
+
+  // make ALL request information visible by scrolling view to it
+  scrollTo( index );
+}
+
+void QgsNetworkLoggerTreeView::contextMenu( QPoint point )
+{
+  const QModelIndex viewModelIndex = indexAt( point );
+  const QModelIndex modelIndex = mProxyModel->mapToSource( viewModelIndex );
+
+  if ( modelIndex.isValid() )
+  {
+    mMenu->clear();
+
+    const QList< QAction * > actions = mLogger->actions( modelIndex, mMenu );
+    mMenu->addActions( actions );
+    if ( !mMenu->actions().empty() )
+    {
+      mMenu->exec( viewport()->mapToGlobal( point ) );
+    }
+  }
+}
+
+void QgsNetworkLoggerTreeView::expandChildren( const QModelIndex &index )
+{
+  if ( !index.isValid() )
+    return;
+
+  const int count = model()->rowCount( index );
+  for ( int i = 0; i < count; ++i )
+  {
+    const QModelIndex childIndex = model()->index( i, 0, index );
+    expandChildren( childIndex );
+  }
+  if ( !isExpanded( index ) )
+    expand( index );
+}
+
+
+//
+// QgsNetworkLoggerPanelWidget
+//
+
+QgsNetworkLoggerPanelWidget::QgsNetworkLoggerPanelWidget( QgsNetworkLogger *logger, QWidget *parent )
+  : QgsDevToolWidget( parent )
+  , mLogger( logger )
+{
+  setupUi( this );
+
+  mTreeView = new QgsNetworkLoggerTreeView( mLogger );
+  verticalLayout->addWidget( mTreeView );
+  mToolbar->setIconSize( QgsGuiUtils::iconSize( true ) );
+
+  mFilterLineEdit->setShowClearButton( true );
+  mFilterLineEdit->setShowSearchIcon( true );
+  mFilterLineEdit->setPlaceholderText( tr( "Filter requests" ) );
+
+  mActionShowTimeouts->setChecked( true );
+  mActionShowSuccessful->setChecked( true );
+  mActionRecord->setChecked( mLogger->isLogging() );
+
+  connect( mFilterLineEdit, &QgsFilterLineEdit::textChanged, mTreeView, &QgsNetworkLoggerTreeView::setFilterString );
+  connect( mActionShowTimeouts, &QAction::toggled, mTreeView, &QgsNetworkLoggerTreeView::setShowTimeouts );
+  connect( mActionShowSuccessful, &QAction::toggled, mTreeView, &QgsNetworkLoggerTreeView::setShowSuccessful );
+  connect( mActionClear, &QAction::triggered, mLogger, &QgsNetworkLogger::clear );
+  connect( mActionRecord, &QAction::toggled, this, [ = ]( bool enabled )
+  {
+    QgsSettings().setValue( QStringLiteral( "logNetworkRequests" ), enabled, QgsSettings::App );
+    mLogger->enableLogging( enabled );
+  } );
+
+  connect( mLogger, &QAbstractItemModel::rowsInserted, this, [ = ]
+  {
+    // we may want to make this optional?
```

Comment:
connect ( mTreeView, &TreeView::scrolled, this, [] () {
  if ( mTreeView->scrollLocation == Bottom )
    mAutoScroll = true;
  else
    mAutoScroll = false;
} );

Reply:
i can't find a reference to scrollLocation?

Reply:
Sorry

```cpp
connect( mTreeView->verticalScrollBar(), &QAbstractSlider::sliderMoved, this, [this] ( int value ) {
  if ( value == mTreeView->verticalScrollBar()->maximum() )
    mAutoScroll = true;
  else
    mAutoScroll = false;
} );
```

Reply:
Thanks!